akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSpec.scala

```diff
@@ -569,39 +569,69 @@ class InterpreterSpec extends AkkaSpec with GraphInterpreterSpecKit {
 
   }
 
-  private[akka] case class Doubler[T]() extends PushPullStage[T, T] {
-    var oneMore: Boolean = false
-    var lastElem: T = _
-
-    override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
-      lastElem = elem
-      oneMore = true
-      ctx.push(elem)
-    }
+  private[akka] final case class Doubler[T]() extends GraphStage[FlowShape[T, T]] {
+    val out: Outlet[T] = Outlet("Doubler.out")
+    val in: Inlet[T] = Inlet("Doubler.in")
+    val _shape = FlowShape(in, out)
```

Comment:
you can provide the impl of shape directly: `override val shape = FlowShape(in,out)` instead of this plus method. Same for the other stage.